File: tools/chaos/runner.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import os
import subprocess
import sys
import time
import uuid
from pathlib import Path
from typing import List

from mttr import fetch_mttr_scorecard, record_events, record_run
from scenarios import SCENARIO_REGISTRY, serialize_results

ROOT = Path(__file__).resolve().parents[2]
HEALTH_CHECK_SCRIPT = ROOT / "proprioception" / "run_health_checks.py"


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Reliability chaos runner for self-healing validation")
    parser.add_argument("--scenarios", nargs="*", default=list(SCENARIO_REGISTRY.keys()))
    parser.add_argument("--mode", default="simulation", choices=["simulation", "scheduled"])
    parser.add_argument("--window-days", type=int, default=30)
    parser.add_argument("--no-regression", action="store_true", help="Skip baseline/post regression checks")
    parser.add_argument("--no-db", action="store_true", help="Do not persist chaos run/events")
    parser.add_argument("--json", action="store_true", help="Print json output")
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Simulation-only run: execute scenarios/regression but skip DB persistence",
    )
    return parser.parse_args()


def run_regression_probe(label: str) -> dict:
    start = time.perf_counter()
    proc = subprocess.run(
        [sys.executable, str(HEALTH_CHECK_SCRIPT), "--dry-run"],
        capture_output=True,
        text=True,
        env=os.environ.copy(),
    )
    elapsed_ms = int((time.perf_counter() - start) * 1000)
    return {
        "label": label,
        "ok": proc.returncode == 0,
        "elapsed_ms": elapsed_ms,
        "stderr": (proc.stderr or "")[:500],
    }


def main() -> int:
    args = parse_args()
    if args.dry_run:
        args.no_db = True
    run_id = str(uuid.uuid4())

    unknown = [s for s in args.scenarios if s not in SCENARIO_REGISTRY]
    if unknown:
        print(f"Unknown scenarios: {', '.join(unknown)}", file=sys.stderr)
        return 2

    regression = []
    if not args.no_regression:
        regression.append(run_regression_probe("pre"))

    results = []
    for name in args.scenarios:
        scenario = SCENARIO_REGISTRY[name]()
        results.append(scenario.run())

    if not args.no_regression:
        regression.append(run_regression_probe("post"))

    serialized = serialize_results(results)
    recovered_count = sum(1 for r in serialized if r["recovered"])
    status = "passed" if recovered_count == len(serialized) and all(r["ok"] for r in regression or [{"ok": True}]) else "failed"

    run_meta = {
        "regression": regression,
        "safe_mode": True,
        "isolation": "temp_files_and_simulated_failures_only",
    }

    if not args.no_db:
        record_run(run_id=run_id, mode=args.mode, scenario_count=len(serialized), status=status, metadata=run_meta)
        record_events(run_id=run_id, events=serialized)

    output = {
        "run_id": run_id,
        "mode": args.mode,
        "status": status,
        "scenarios": serialized,
        "regression": regression,
        "mttr_scorecard": fetch_mttr_scorecard(window_days=args.window_days) if not args.no_db else None,
    }

    if args.json:
        print(json.dumps(output, indent=2))
    else:
        print(f"Chaos run {run_id}: {status} ({len(serialized)} scenarios)")
        for s in serialized:
            print(f" - {s['name']}: detected={s['detected']} recovered={s['recovered']} recovery_ms={s['recovery_ms']}")

    return 0 if status == "passed" else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/chaos/runner.ts
#!/usr/bin/env npx tsx

import { spawnSync } from "node:child_process";
import path from "node:path";
import { randomUUID } from "node:crypto";
import { fetchMttrScorecard, recordEvents, recordRun } from "./mttr.js";
import { serializeResults } from "./scenarios/base.js";
import { CronFailureScenario } from "./scenarios/cron_failure.js";
import { DbConnectionIssueScenario } from "./scenarios/db_connection_issue.js";
import { HeartbeatMissScenario } from "./scenarios/heartbeat_miss.js";
import { MemoryCorruptionScenario } from "./scenarios/memory_corruption.js";
import { ToolUnavailabilityScenario } from "./scenarios/tool_unavailability.js";
import { repoRoot } from "../lib/paths.js";

type Args = {
  scenarios: string[];
  mode: "simulation" | "scheduled";
  windowDays: number;
  noRegression: boolean;
  noDb: boolean;
  json: boolean;
  dryRun: boolean;
};

const SCENARIO_REGISTRY: Record<string, new () => { run(): Promise<any> }> = {
  tool_unavailability: ToolUnavailabilityScenario,
  cron_failure: CronFailureScenario,
  db_connection_issue: DbConnectionIssueScenario,
  memory_corruption: MemoryCorruptionScenario,
  heartbeat_miss: HeartbeatMissScenario,
};

const HEALTH_CHECK_SCRIPT = path.join(repoRoot(), "proprioception", "run_health_checks.py");

function parseArgs(argv: string[]): Args {
  const args: Args = {
    scenarios: Object.keys(SCENARIO_REGISTRY),
    mode: "simulation",
    windowDays: 30,
    noRegression: false,
    noDb: false,
    json: false,
    dryRun: false,
  };

  for (let i = 0; i < argv.length; i += 1) {
    const a = argv[i];
    if (a === "--scenarios") {
      const vals: string[] = [];
      let j = i + 1;
      while (j < argv.length && !argv[j].startsWith("--")) {
        vals.push(argv[j]);
        j += 1;
      }
      args.scenarios = vals.length ? vals : [];
      i = j - 1;
    } else if (a === "--mode") {
      args.mode = (argv[++i] as Args["mode"]) ?? "simulation";
    } else if (a === "--window-days") {
      args.windowDays = Number.parseInt(argv[++i] ?? "30", 10);
    } else if (a === "--no-regression") {
      args.noRegression = true;
    } else if (a === "--no-db") {
      args.noDb = true;
    } else if (a === "--json") {
      args.json = true;
    } else if (a === "--dry-run") {
      args.dryRun = true;
    }
  }

  return args;
}

function runRegressionProbe(label: string): Record<string, unknown> {
  const start = performance.now();
  const proc = spawnSync(process.execPath, [HEALTH_CHECK_SCRIPT, "--dry-run"], {
    encoding: "utf8",
    env: { ...process.env },
  });
  const elapsedMs = Math.trunc(performance.now() - start);
  return {
    label,
    ok: proc.status === 0,
    elapsed_ms: elapsedMs,
    stderr: (proc.stderr || "").slice(0, 500),
  };
}

async function main(): Promise<number> {
  const args = parseArgs(process.argv.slice(2));
  if (args.dryRun) args.noDb = true;
  const runId = randomUUID();

  const unknown = args.scenarios.filter((s) => !(s in SCENARIO_REGISTRY));
  if (unknown.length) {
    console.error(`Unknown scenarios: ${unknown.join(", ")}`);
    return 2;
  }

  const regression: Array<Record<string, unknown>> = [];
  if (!args.noRegression) regression.push(runRegressionProbe("pre"));

  const results = [];
  for (const name of args.scenarios) {
    const scenario = new SCENARIO_REGISTRY[name]();
    results.push(await scenario.run());
  }

  if (!args.noRegression) regression.push(runRegressionProbe("post"));

  const serialized = serializeResults(results as any[]);
  const recoveredCount = serialized.filter((r) => r.recovered).length;
  const status = recoveredCount === serialized.length && (regression.length ? regression.every((r) => r.ok === true) : true) ? "passed" : "failed";

  const runMeta = {
    regression,
    safe_mode: true,
    isolation: "temp_files_and_simulated_failures_only",
  };

  if (!args.noDb) {
    recordRun(runId, args.mode, serialized.length, status, runMeta);
    recordEvents(runId, serialized as any[]);
  }

  const output = {
    run_id: runId,
    mode: args.mode,
    status,
    scenarios: serialized,
    regression,
    mttr_scorecard: !args.noDb ? fetchMttrScorecard(args.windowDays) : null,
  };

  if (args.json) {
    console.log(JSON.stringify(output, null, 2));
  } else {
    console.log(`Chaos run ${runId}: ${status} (${serialized.length} scenarios)`);
    for (const s of serialized) {
      console.log(` - ${s.name}: detected=${s.detected} recovered=${s.recovered} recovery_ms=${s.recovery_ms}`);
    }
  }

  return status === "passed" ? 0 : 1;
}

main().then((code) => process.exit(code));
